### PR TITLE
Rename signature fix

### LINKED ISSUE #17963: API break with DataFrame.rename
Posted in https://github.com/pandas-dev/pandas/pull/17800#issuecomment-338989093, but making a new issue to make sure it's not lost and that everone is aware, and to check if we're comfortable with this:

```python
df = pd.DataFrame(columns=['A', 'B'])
df.rename(None, str.lower)
```

In previous versions that meant `.rename(index=None, columns=str.lower)`, but now it's `.rename(mapper=None, index=str.lower)`. I think this is fundamentally ambiguous correct? This is breaking dask, which used positional arguments, but I suspect it'll break more code.

cc @jreback @jorisvandenbossche

### PR DESCRIPTION
**Pull request by TomAugspurger**
- [ ] closes #17963
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Sample usage:

```
In [3]: df = pd.DataFrame(index=['a', 'b'], columns=['A', 'B'])

In [4]: df.reindex(['a'], ['A'])
/Users/taugspurger/.virtualenvs/pandas-dev/bin/ipython:1: UserWarning: Intepreting call
        '.reindex(a, b)' as
        '.reindex(index=a, columns=b)'.
Use named arguments to remove any ambiguity.
  #!/Users/taugspurger/Envs/pandas-dev/bin/python3.6
Out[4]:
     A
a  NaN

In [5]: df.reindex(None, ['A'])
/Users/taugspurger/.virtualenvs/pandas-dev/bin/ipython:1: UserWarning: Intepreting call
        '.reindex(a, b)' as
        '.reindex(index=a, columns=b)'.
Use named arguments to remove any ambiguity.
  #!/Users/taugspurger/Envs/pandas-dev/bin/python3.6
Out[5]:
     A
a  NaN
b  NaN
```

I'd like to simplify this further if possible, things go messy :/ 
I'd also like to fix the function signature, which is doable on python3 at least.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Added signature re-writing for python3 only. AFAICT, there's not a simple way to do that for both python 2 and 3.

**Comment by jorisvandenbossche** [comment]
In principle this is fine for me (anyhow, the solution to interpret it as positional index=, columns= but warning is good), but have the feeling it is more complex as needed? 
In the case the user passes positional arguments `index, columns`, this is captured as `labels=index, index=columns`, but specifying both `labels` as `index` is not valid? So normally we should raise for that case, but can't we catch this situation, reinterprete and do the warning as in this PR? Or am I missing something?

**Comment by TomAugspurger** [comment]
I’ll see if that works, it would be more maintainable. One drawback of that is having to treat a user explicitly passing labels=foo, index=bar the same as just passing two positional args. Currently we would raise on that, instead we would issue a warning.

________________________________

**Comment by jorisvandenbossche** [comment]
Ah, yes, that's a case I didn't think of. Personally, I would go for accepting that minor disadvantage to have the simpler code. The question then maybe is if we would make the warning stronger in mentioning that this will be changed in the future (actually deprecate instead of a UserWarning), so that in the future we can actually raise for the labels=foo, index=bar case

File: pandas/util/_decorators.py [hunk 1]
```diff
@@ -122,6 +123,31 @@ def wrapper(*args, **kwargs):
...
+                inspect.Parameter('columns', kind, default=None),
+                inspect.Parameter('axis', kind, default=None),
+            ]
+
+            for pname, default in extra_params:
+                params.append(inspect.Parameter(pname, kind, default=default))
+
+            sig = inspect.Signature(params)
+
+            func.__signature__ = sig
```
Comment (jreback) [inline pandas/util/_decorators.py]: cool that u can do this

File: pandas/core/panel.py [hunk 2]
```diff
@@ -1199,20 +1199,31 @@ def _wrap_result(self, result, axis):
...
-                items=None, major_axis=None, minor_axis=None,
-                axis=None, **kwargs):
-        major_axis = (major_axis if major_axis is not None else
-                      kwargs.pop('major', None))
-        minor_axis = (minor_axis if minor_axis is not None else
-                      kwargs.pop('minor', None))
-        axes = self._validate_axis_style_args(
-            labels, 'labels', axes=[items, major_axis, minor_axis],
-            axis=axis, method_name='reindex')
+    def reindex(self, *args, **kwargs):
```
Comment (jreback) [inline pandas/core/panel.py]: it might actually make this more complicated, but would be ok with leaving panel/panel4d as is.

Comment (TomAugspurger) [reply pandas/core/panel.py]: I briefly tried that, but had to add several `if obj.ndim >= 3` inside internals. Got a bit messy.

**Comment by TomAugspurger** [comment]
> Ah, yes, that's a case I didn't think of. Personally, I would go for accepting that minor disadvantage to have the simpler code

Hmm I'm having trouble making the simple way (no `*args, **kwargs)` work. IIUC, then all three of these have to be handled the same way (this is master):

```python
In [15]: df.rename(None, str.lower)  # <----- This is the problem one
Out[15]:
     A    B
a  NaN  NaN
b  NaN  NaN

In [16]: df.rename(None, index=str.lower)
Out[16]:
     A    B
a  NaN  NaN
b  NaN  NaN

In [17]: df.rename(mapper=None, index=str.lower)
Out[17]:
     A    B
a  NaN  NaN
b  NaN  NaN
```

We want to treat `[15]` differently, since that is (presumably) using the old index/columns style.

**Comment by TomAugspurger** [comment]
> The question then maybe is if we would make the warning stronger in mentioning that this will be changed in the future

Once we're python3 only, we could using keyword-only arguments I think:

```python
def rename(mapper=None, *, index=None, columns=None, axis=None, ...):
```

I should be able to emit a warning whenever `len(args) > 1`, and in the future we can have a regular signature and remove all the hacks in this PR.

**Comment by jorisvandenbossche** [comment]
> IUC, then all three of these have to be handled the same way (this is master):

I personally don't mind doing that (just all treating them as positional to deal with [15]). There will be a warning in any case, so people doing one of those three cases will typically change their code to not have the warning anymore. 
In the case of `df.rename(mapper=None, index=str.lower)` the warning will not be fully accurate, but then, this is really a very strange thing to do.

**Comment by TomAugspurger** [comment]
I'd be fine with raising on `In[17]`, since if they passed anything other than `None` we would already raise. But I think `In[16]` is perfectly fine. We would be reinterpreting that as `df.rename(index=None, columns=str.lower)`, or just `.rename(columns=str.lower)`.

Let me try using different defaults other than None to see if that helps at all...

**Comment by jorisvandenbossche** [comment]
@TomAugspurger if it is too complicated, I am also fine with this though! It's working, and we can later on remove the hacks again.

**Comment by TomAugspurger** [comment]
FWIW, I find this version of `_validate_axis_style_kwargs` more understandable, but presumably because it's more recent :) My last commit add a docstring / examples, so hopefully that helps.

**Comment by jorisvandenbossche** [comment]
To be clear, I mainly meant the signature hacking, I didn't look at the changes in `_validate_axis_style_kwargs` (since it moved it is a bit difficult to see what changed)

**Comment by TomAugspurger** [comment]
Ah, yeah. FWIW, I'm 51/49 on whether overwriting the signature is a good thing or not. It's going to confuse static type checkers anyway. I'm OK with it for now since it'll be removed once we're Python3 only.

**Comment by TomAugspurger** [comment]
OK, this should be all good.

File: pandas/core/generic.py [hunk 3]
```diff
@@ -2822,6 +2777,98 @@ def sort_index(self, axis=0, level=None, ascending=True, inplace=False,
...
+
+            msg = ("Intepreting call\n\t'.{method_name}(a, b)' as "
+                   "\n\t'.{method_name}(index=a, columns=b)'.\nUse named "
+                   "arguments to remove any ambiguity. In the future, using "
+                   "positional arguments for 'index' or 'columns' will raise "
+                   " a 'TypeError'.")
+            warnings.warn(msg.format(method_name=method_name,), FutureWarning,
+                          stacklevel=4)
+            out[self._AXIS_NAMES[0]] = args[0]
+            out[self._AXIS_NAMES[1]] = args[1]
```
Comment (jreback) [inline pandas/core/generic.py]: would really love to move this *out* of generic (and just pass ``self`` into a static function, maybe into pandas.compat.validations or something).

File: pandas/core/generic.py [hunk 4]
```diff
@@ -2822,6 +2777,98 @@ def sort_index(self, axis=0, level=None, ascending=True, inplace=False,
...
+
+        if len(args) == 0:
+            pass  # It's up to the function to decide if this is valid
+        elif len(args) == 1:
+            axis = self._get_axis_name(kwargs.get('axis', 0))
+            out[axis] = args[0]
+        elif len(args) == 2:
+            if 'axis' in kwargs:
+                # Unambiguously wrong
+                msg = "Cannot specify both {} and any of 'index' or 'columns'"
```
Comment (jorisvandenbossche) [inline pandas/core/generic.py]: Shouldn't this be "Cannot specify both axis and index / columns" ?(as you check that axis is specified)

File: pandas/core/generic.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/core/generic.py]: This is only used by reindex/rename, and now is sits next to those functions. IMO it is easier to keep it that way.

File: pandas/core/generic.py (hunk 4 above)
Comment (jorisvandenbossche) [reply pandas/core/generic.py]: (and maybe ' ' around {})

**Review by jorisvandenbossche — COMMENTED** [review]
Now also looked at the rewritten validate_axis_args method. Looks good!

**Comment by TomAugspurger** [comment]
@jreback are you OK with it in generic? I use `self` to figure out whether we're a Panel or a DataFrame, and it's only used on those two methods, so I think it's OK where it is.

**Comment by jreback** [comment]
> @jreback are you OK with it in generic? I use self to figure out whether we're a Panel or a DataFrame, and it's only used on those two methods, so I think it's OK where it is.

I don't really like it, because this is really validation code.

If you could move it would be much better (and forr sure pass in ``self`` as the arg)

**Comment by TomAugspurger** [comment]
Moved to `utils/_validators.py`.

**Review by jreback — APPROVED** [review]
lgtm

**Pull request merged**